fcrepo-http-api/src/main/java/org/fcrepo/http/api/PathLockManager.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed to DuraSpace under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership.
+ *
+ * DuraSpace licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.fcrepo.http.api;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Semaphore;
+
+import org.slf4j.Logger;
+
+/**
+ * A class that serves as a pool of locks to guarantee synchronized
+ * access based on the String representation of paths.  Because there
+ * may be an extremely high number of paths in the repository, the main
+ * function of this class is to ensure that for a given path only one
+ * semaphore exists, and that that semaphore exists only as long as a
+ * caller wishes to access that path.
+ *
+ * The correct usage pattern is to invoke getSemaphoreForPath (a blocking
+ * method), perform some mutations to the resource at that path, then
+ * invoke release on that semaphore.
+ *
+ * @author md5wz
+ */
+public class PathLockManager {
+
+    private static final Logger LOGGER = getLogger(PathLockManager.class);
+
+    private Map<String, TransientSemaphore> pool = new HashMap<String, TransientSemaphore>();
+
+    /**
+     * A specialized construct for the purpose of allowing mutex controls (blocking)
+     * on arbitrary keys that only exist in memory as long as there is activity against
+     * that particular key.  The creation of these objects should be handled entirely
+     * by the method getSemaphoreForPath().
+     *
+     * @author md5wz
+     */
+    public class TransientSemaphore {
+
+        private String key;
+
+        private Semaphore wrappedSemaphore;
+
+        private int count;
+
+        /**
+         * Default constructor.
+         * @param key the locking key
+         */
+        public TransientSemaphore(final String key) {
```

Comment:
Only for that semaphore. Not for the whole pool. So get the semaphore in question from the map and then synchronize on it. But see my next comment.